### ISSUE TITLE
Custom source not working

### ISSUE THREAD

**Issue by kristijanhusak** [Shougo/deoplete.nvim#815]
# Problems summary
Some time ago i created custom source https://github.com/kristijanhusak/deoplete-phpactor. It was working at that time, but now it's not working. I validated that the source is properly called and the results are there.

## Expected
To get the results in autocomplete

## Environment Information

 * deoplete version (SHA1): Latest

 * OS: Linux mint 19

 * neovim/Vim `:version` output:
```
NVIM v0.3.2-dev
Build type: RelWithDebInfo
Lua 5.1
Compilation: /usr/bin/cc -g -O2 -fdebug-prefix-map=/build/neovim-18_sLu/neovim-0.3.0~ubuntu1+git201807252033-9c0c5e8-479a1d0=. -fstack-protector-strong -Wformat -Werror=format-security -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -DMIN_LOG_LEVEL=3 -Og -g -Wall
-Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -Wno-array-bounds -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/bui
ld/neovim-18_sLu/neovim-0.3.0~ubuntu1+git201807252033-9c0c5e8-479a1d0/build/config -I/build/neovim-18_sLu/neovim-0.3.0~ubuntu1+git201807252033-9c0c5e8-479a1d0/src -I/build/neovim-18_sLu/neovim-0.3.0~ubuntu1+git201807252033-9c0c5e8-479a1d0/.deps/usr/include -I/usr/include -I
/build/neovim-18_sLu/neovim-0.3.0~ubuntu1+git201807252033-9c0c5e8-479a1d0/build/src/nvim/auto -I/build/neovim-18_sLu/neovim-0.3.0~ubuntu1+git201807252033-9c0c5e8-479a1d0/build/include
Compiled by buildd@lcy01-amd64-029

Features: +acl +iconv +jemalloc +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"
```

 * `:checkhealth` or `:CheckHealth` result(neovim only):
```
health#deoplete#check
========================================================================
## deoplete.nvim
  - OK: has("nvim") was successful
  - OK: exists("v:t_list") was successful
  - OK: has("timers") was successful
  - OK: has("python3") was successful
  - INFO: If you're still having problems, try the following commands:
    $ export NVIM_PYTHON_LOG_FILE=/tmp/log
    $ export NVIM_PYTHON_LOG_LEVEL=DEBUG
    $ nvim
    $ cat /tmp/log_{PID}
    and then create an issue on github

health#nvim#check
========================================================================
## Configuration
  - OK: no issues found

## Performance
  - OK: Build type: RelWithDebInfo

## Remote Plugins
  - OK: Up to date

## terminal
  - INFO: key_backspace (kbs) terminfo entry: key_backspace=\177
  - INFO: key_dc (kdch1) terminfo entry: key_dc=\E[3~
  - INFO: $VTE_VERSION='5202'
  - INFO: $COLORTERM='truecolor'

## tmux
  - OK: escape-time: 0ms
  - INFO: $TERM: xterm-256color-italic
  - ERROR: $TERM should be "screen-256color" or "tmux-256color" in tmux. Colors might look wrong.
    - ADVICE:
      - Set default-terminal in ~/.tmux.conf:
          set-option -g default-terminal "screen-256color"
      - https://github.com/neovim/neovim/wiki/FAQ

health#provider#check
========================================================================
## Clipboard (optional)
  - OK: Clipboard tool found: xclip

## Python 2 provider (optional)
  - INFO: `g:python_host_prog` is not set.  Searching for python2 in the environment.
  - INFO: Executable: /usr/bin/python2
  - INFO: Python2 version: 2.7.15
  - INFO: python2-neovim version: 0.2.4 (outdated; from /usr/local/lib/python2.7/dist-packages/neovim)
  - WARNING: Latest python2-neovim is NOT installed: 0.2.6

## Python 3 provider (optional)
  - INFO: `g:python3_host_prog` is not set.  Searching for python3 in the environment.
  - INFO: Executable: /usr/bin/python3
  - INFO: Python3 version: 3.6.5
  - INFO: python3-neovim version: 0.2.6
  - OK: Latest python3-neovim is installed: 0.2.6

## Ruby provider (optional)
  - INFO: Ruby: ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux-gnu]
  - WARNING: `neovim-ruby-host` not found.
    - ADVICE:
      - Run `gem install neovim` to ensure the neovim RubyGem is installed.
      - Run `gem environment` to ensure the gem bin directory is in $PATH.
      - If you are using rvm/rbenv/chruby, try "rehashing".
      - See :help |g:ruby_host_prog| for non-standard gem installations.

## Node.js provider (optional)
  - INFO: Node.js: v8.7.0
  - INFO: Neovim node.js host: /home/kristijan/.nvm/versions/node/v8.7.0/lib/node_modules/neovim/bin/cli.js
  - WARNING: Package "neovim" is out-of-date. Installed: 3.5.2, latest: 4.2.1
    - ADVICE:
      - Run in shell: npm install -g neovim
```

## Provide a minimal init.vim/vimrc with less than 50 lines (Required!)

```vim
set nocompatible hidden laststatus=2

if !filereadable('/tmp/plug.vim')
  silent !curl --insecure -fLo /tmp/plug.vim
    \ https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
endif

source /tmp/plug.vim
call plug#begin('/tmp/plugged')
Plug 'Shougo/deoplete.nvim', { 'do': ':UpdateRemotePlugins' }
Plug 'kristijanhusak/deoplete-phpactor'
Plug 'phpactor/phpactor', { 'do': 'composer install' }
call plug#end()

autocmd VimEnter * PlugClean! | PlugUpdate --sync
let g:deoplete#enable_at_startup = 1
```

## How to reproduce the problem from neovim/Vim startup (Required!)

 1. Make sure to have php7+ and composer installed
 2. Clone this repo https://github.com/kristijanhusak/laravel-form-builder
 3. do `composer install` in it
 4. open file with above vimrc, go to file `src/Kris/LaravelFormBuilder/Form.php`
 5. Try autocompleting in any method something like `$this->formHelper->get`

I validated that the candidates are populated from the phpactor rpc (did an ugly self.print_error before the return in gather_candidates), but for some reason they are not returned to autocomplete.

**Comment by Shougo** [comment]
I have tested it.
It works.   But you need to wait.
It takes time.

**Comment by kristijanhusak** [comment]
I tried waiting, but still nothing.Is there anything missing in my config maybe?

**Comment by Shougo** [comment]
My config is here.

```vim
set rtp+=~/work/deoplete.nvim/
set rtp+=~/work/nvim-yarp
set rtp+=~/work/vim-hug-neovim-rpc/
set rtp+=~/src/deoplete-phpactor/
set rtp+=~/src/phpactor/

let g:deoplete#enable_at_startup = 1
let g:deoplete#enable_refresh_always = 1
```

**Comment by Shougo** [comment]
Unfortuntely, I ignore not reproduced issue.

**Comment by Shougo** [comment]
The patch may be needed.

```diff
diff --git a/autoload/deoplete/mapping.vim b/autoload/deoplete/mapping.vim
index 7071a94..28b5108 100644
--- a/autoload/deoplete/mapping.vim
+++ b/autoload/deoplete/mapping.vim
@@ -10,7 +10,8 @@ function! deoplete#mapping#_init() abort
 endfunction

 function! deoplete#mapping#_dummy_complete() abort
-  return "\<C-r>=deoplete#mapping#_complete()\<CR>"
+  return (pumvisible() ? "\<C-e>" : "")
+        \ . "\<C-r>=deoplete#mapping#_complete()\<CR>"
 endfunction
 function! deoplete#mapping#_completefunc(findstart, base) abort
   if a:findstart
```

> @Shougo i understand that. Can you just point me where i should debug this?

Why don't use `git bisect`?

**Comment by kristijanhusak** [comment]
Patch didn't work out for me. I'm also using your vimrc that you provided.
I didn't investigate phpactor because it does return proper results in python.
Can i maybe make this source more async or something? Or maybe to investigate on a level where it should come to complete popup?
I also tried using manual complete, and it didn't return me those. I got all buffer, member and other stuff properly.
